### PR TITLE
Instrumental variables

### PR DESCRIPTION
**Pull request by jmafoster1**
Preliminary implementation of instrumental variable estimation.

### CODE REVIEW

File: causal_testing/specification/causal_dag.py [hunk 1]
```diff
@@ -138,6 +138,36 @@ def __init__(self, dot_path: str = None, **attr):
...
+            raise ValueError(f"Instrument {instrument} is not associated with treatment {treatment} in the DAG")
+
+        # (ii) Instrument does not affect outcome except through its potential effect on treatment
+        if not all([treatment in path for path in nx.all_simple_paths(self.graph, source=instrument, target=outcome)]):
+            raise ValueError(
+                f"Instrument {instrument} affects the outcome {outcome} other than through the treatment {treatment}"
+            )
+
+        # (iii) Instrument and outcome do not share causes
+        if any(
```
Comment (cwild-UoS) [inline causal_testing/specification/causal_dag.py:159]: I think the logic could be easier to read if written like this:
```
if any(
    list(nx.all_simple_paths(self.graph, source=cause, target=instrument))
    and list(nx.all_simple_paths(self.graph, source=cause, target=outcome))
    for cause in self.graph.nodes
):
    raise ValueError(f"Instrument {instrument} and outcome {outcome} share common causes")
```

**Review by cwild-UoS — APPROVED** [review]
Looks good to me!

Thanks for catching so many awkward remaining tuples

File: causal_testing/testing/estimators.py [hunk 2]
```diff
@@ -133,20 +133,20 @@ def _run_logistic_regression(self, data) -> RegressionResultsWrapper:
         """
         # 1. Reduce dataframe to contain only the necessary columns
         reduced_df = data.copy()
-        necessary_cols = list(self.treatment) + list(self.adjustment_set) + list(self.outcome)
+        necessary_cols = [self.treatment] + list(self.adjustment_set) + [self.outcome]
```
[the remark this answers is not in the document]
Comment (jmafoster1) [reply causal_testing/testing/estimators.py:136]: Adjustment set really is a set (not just a singleton) so we need to have it be a list. If we did `[self.adjustment_set]`, we'd end up with something that looked like `[{"X"}]` which isn't what we want.

File: causal_testing/testing/estimators.py [hunk 3]
```diff
@@ -536,12 +536,65 @@ def _run_linear_regression(self) -> RegressionResultsWrapper:
...
+        Estimate the linear regression coefficient of the treatment on the outcome.
+        """
+        # Estimate the total effect of instrument I on outcome Y = abI + c1
+        ab = sm.OLS(self.df[self.outcome], self.df[[self.instrument]]).fit().params[self.instrument]
+
+        # Estimate the direct effect of instrument I on treatment X = aI + c1
+        a = sm.OLS(self.df[self.treatment], self.df[[self.instrument]]).fit().params[self.instrument]
+
+        # Estimate the coefficient of I on X by cancelling
+        return ab / a
```
[the remark this answers is not in the document]
Comment (jmafoster1) [reply causal_testing/testing/estimators.py:592]: I doubt it'd ever actually really be 0.0, especially if the user is expecting an effect to be there

File: causal_testing/testing/estimators.py [hunk 4]
```diff
@@ -536,12 +536,65 @@ def _run_linear_regression(self) -> RegressionResultsWrapper:
...
+
+    def estimate_ate(self):
+        return (self.treatment_value - self.control_value) * self.estimate_coefficient(), (None, None)
```
[the remark this answers is not in the document]
Comment (jmafoster1) [reply causal_testing/testing/estimators.py:595]: Type checking for other methods expecting a tuple. I eventually intend to implement proper confidence intervals

File: causal_testing/specification/causal_dag.py (hunk 1 above)
Comment (jmafoster1) [reply causal_testing/specification/causal_dag.py:159]: Fair enough. I'll fix

**Pull request merged**